mesonbuild/scripts/python_info.py

```diff
@@ -58,10 +58,15 @@ def get_install_paths():
     return paths, install_paths
 
 def links_against_libpython():
-    from distutils.core import Distribution, Extension
-    cmd = Distribution().get_command_obj('build_ext')
-    cmd.ensure_finalized()
-    return bool(cmd.get_libraries(Extension('dummy', [])))
+    # on versions supporting python-embed.pc, this is the non-embed lib
+    if sys.version_info >= (3, 8):
+        variables = sysconfig.get_config_vars()
+        return bool(variables.get('LIBPYTHON', 'yes'))
```

Comment:
Just nitpicking, but 
```python
bool(variables.get('LIBPYTHON', True))
```
seems more idiomatic to me.

Reply:
Unfortunately this does not work on PyPy. On PyPy 3.9 and 3.10 `sysconfig.get_config_vars()` does not contain a `LIBPYTHON` key, thus `.get()` returns the default value also on Linux and macOS. Either the fallback to `distutils` is applied for Python versions up to 3.10, or it is also made conditional on the interpreter being PyPy.

Reply:
The current code assumes that if the `LIBPYTHON` variable is not present, the code is being executed on Windows, and thus libpython needs to be linked. Therefore, another possibility would be to do not make this assumption and make the check explicit by replacing
```python
return bool(variables.get('LIBPYTHON', 'yes'))
``` 
with 
```python
return sys.platform == 'win32' or bool(variables.get('LIBPYTHON'))
```

Reply:
Do we know what PyPy did across various different historic versions?

Reply:
I've tested only pypy3.9 and pypy3.10

Reply:
I did some archeology. To the best of my knowledge, PyPy never exposed a `LIBPYTHON` variable via `sysconfig.get_config_var()`. On PyPy `_sysconfigdata` is a regular python module, it is not generated, thus verifying this is not very complicated.

Reply:
I am more hoping to know what PyPy expected in terms of "should you link to the PyPy library".

Reply:
I misunderstood. Let me see how hard is to figure this out.

Reply:
Note that the current approach has the habit of erring on the side of "caution" / overlinking.

Reply:
I investigated some more. From what I can tell, PyPy never requires and never required to link to a libpython equivalent on platforms other than Windows.

Reply:
https://foss.heptapod.net/pypy/pypy/-/commit/3c1451fcce7cf15221a10cd443b8d195cdf47f27#052bad03bac8273f96b621028a7d3cf7beaef554_709_700

Looks like, if you install PyPy 7.3.12 (Python 2.7.18) you get this content in distutils:

```
    def get_libraries (self, ext):
        """Return the list of libraries to link against when building a
        shared extension.  On most platforms, this is just 'ext.libraries';
        on Windows and OS/2, we add the Python library (eg. python20.dll).
        """
        # For PyPy, we must not add any such Python library, on any platform
        if "__pypy__" in sys.builtin_module_names:
            return ext.libraries
```

It was more or less reverted to stock CPython semantics in the py3.3 branch back in 2014, and continued that way until I guess CPython switched over.

@mattip, what's the expected semantics for PyPy?

Reply:
Uhh, I think "never specify libpython in the command line link arguments" is correct for PyPy. For windows, there is a pragma hidden in the header files that forces linking with an import library, and linux/macos has never needed to link with the libpython.so. Does that thesis hold up in tests?

Reply:
I am open to adding sysconfig values that would help here. What do you need that PyPy does not provide?

Reply:
There are complications to that. For example:
- Android is more or less a Linux, but the way they have it set up, libpython.so is RTLD_LOCAL and thus not available to built extensions.
- the Windows pragma is dark sorcery and also unfortunately doesn't reliably work, so we pass `/NODEFAULTLIB:pythonXXXXXX` to deactivate the pragma, and then link to a library that has a different name vs the pragma. I also don't know how the pragma is supposed to work at all without passing the library directory, which we generally don't do as we resolve DLLs to filepaths

If PyPy has a consistent "only do it on Windows" policy for extension modules then we can work with that. :)

> I am open to adding sysconfig values that would help here. What do you need that PyPy does not provide?

The internal implementation of CPython's sysconfig is that `sysconfig.get_config_var('LIBPYTHON')` contains the Makefile variable that gets formatted into pkg-config files and `python-config --libs` if the autoconf-based build detects that you need to specify libpython in the command line. Since Windows doesn't use the autoconf-based build, that currently means a) cygwin, b) android, and also I'm pretty sure unofficially c) msys2

So adding LIBPYTHON as equal to `''` on linux/macOS would be useful, I think. It cannot hurt to have it equal to `'libpypy3.10-c.dll'` or whatever on Windows too. I am not sure cygwin/android are supported anyway?

Reply:
PyPy does not support officially cygwin/android. I see on windows, CPython returns None for `LIBPYTHON` on v3.11.5, but I will change it on PyPy as you requested.

Reply:
> If PyPy has a consistent "only do it on Windows" policy for extension modules

It is not PyPy: it is MSVC that wants to resolve all linker symbols at link time, even if making a shared object.

Reply:
> I see on windows, CPython returns None for `LIBPYTHON` on v3.11.5, but I will change it on PyPy as you requested.

This is actually the CPython behavior meson expects: `LIBPYTHON` should be undefined on Windows and set to `''` on macOS and Linux. As PyPy does not generate the data that is returned by the `sysconfig` module dynamically, I don't know whether it makes sense to generate the right value at runtime in `_sysconfigdata.

Reply:
So how are consumers meant to know the name of the import lib and its path on windows? The discussion at the top of the [sysconfig rewrite](https://github.com/python/cpython/issues/103482) suggests 
> Is a shared libpython available? What's its name? Location?
>  - definition: Yes if built with --enable-shared
>  - currently:
>    - (available?) bool(sysconfig.get_config_var('LDLIBRARY'))
>    - (name?) sysconfig.get_config_var('LDLIBRARY')
>    - (location?) sysconfig.get_config_var('LIBDIR')

but I don't think reality concurs with the "currently" status. Or maybe I don't understand the intent.

Reply:
Please note that the variable names in the text you quote from the CPython PR and the one used here are different, thus I'm not sure how that fits here. A bit of magic goes into obtaining the library name to use and its location. This is because there isn't a reliable way to get that information that works on all architectures across all versions of Python supported by Meson. This is unfortunate and a bit more complex that what is inferred from the CPython PR you link to.

I don't think that the quirks of how the information required for building extension modules can be retrieved from a CPython installation are nice or the best way things could be, but having all Python implementation implement the same interface helps the consumers of such information. At the same time, there is no fixing for the already released PyPy versions, thus Meson will need to find a way to work with what is out there.

Reply:
> So how are consumers meant to know the name of the import lib and its path on windows? The discussion at the top of the sysconfig rewrite suggests

You are quite right, CPython doesn't really expose this information at all for Visual Studio solutions file builds and that is partially because those builds are a *really* odd duck in a whole bunch of ways and partially because there isn't really an official way to get this information at all.

The sysconfig variables we're discussing here are the best we have and they don't really work at all on Windows...

... which is why I suggested that once we are adding stuff manually to PyPy's sysconfigdata, it might be good to go the extra mile and provide this information even on Windows, where CPython doesn't provide it. :D

Reply:
> but I don't think reality concurs with the "currently" status. Or maybe I don't understand the intent.

It tries to show how people generally get that value at the moment, not to suggest how they should do it. It is also very Posix-centric as that's my background and that issue was not meant to be an extensive document — just help to paint the picture. [PEP 720](https://peps.python.org/pep-0720/) OTOH tries to do that, but it still lacks a lot of information.

This PR is a good example of why the current way is problematic, which is what we want to fix.

> So how are consumers meant to know the name of the import lib and its path on windows?

I don't think there is a portable/reliable way, unfortunately.

https://github.com/python/cpython/pull/110049 will help with that, but only on >=3.13.

Reply:
One of the advantanges of PyPy is that I can make changes that will be available in a nightly build of python3.9/3.10, and these can be easily tried out in CI. So if the direction in PR 10049 looks correct, I can add that to PyPy so that, at least on PyPy, meson could get the values from the installed python.